modules/workergroup/nodepools.tf

```diff
@@ -0,0 +1,115 @@
+        max_pods_per_node = min(max(var.max_pods_per_node, 1), 110)
+        # pick the 1st pod nsg here until https://github.com/oracle/terraform-provider-oci/issues/1662 is clarified and resolved
+        pod_nsg_ids    = slice(each.value.pod_nsg_ids, 0, min(1, length(each.value.pod_nsg_ids)))
+        pod_subnet_ids = tolist([coalesce(each.value.pod_subnet_id, var.subnet_id)])
+      }
+    }
+
+    defined_tags = merge(
+      local.defined_tags,
+      lookup(each.value, "defined_tags", {}),
+    )
+    freeform_tags = merge(local.freeform_tags, contains(keys(each.value), "freeform_tags") ? each.value.freeform_tags : { worker_group = each.key })
+  }
+
+  node_metadata = {
+    apiserver_host           = var.apiserver_private_host
+    kubedns_svc_ip           = var.cluster_dns
+    oke-kubeproxy-proxy-mode = var.kubeproxy_mode
+    user_data                = data.cloudinit_config.worker_once.rendered
+  }
+
+  node_shape = each.value.shape
+  dynamic "node_shape_config" {
+    for_each = length(regexall("Flex", each.value.shape)) > 0 ? [1] : []
+    content {
+      ocpus = each.value.ocpus
+      memory_in_gbs = ( # If > 64GB memory/core, correct input to exactly 64GB memory/core
+        (each.value.memory / each.value.ocpus) > 64 ? each.value.ocpus * 64 : each.value.memory
+      )
+    }
+  }
+
+  dynamic "node_source_details" {
+    for_each = length(local.parsed_images) > 0 ? [1] : []
+    content {
+      source_type             = local.node_pool_images[0].source_type
+      boot_volume_size_in_gbs = each.value.boot_volume_size
+      image_id = (each.value.image_type == "custom" ? each.value.image_id
+        : element(tolist(setintersection([
+          lookup(local.image_ids, each.value.image_type, null),
+          length(regexall("GPU", each.value.shape)) > 0 ? local.image_ids.gpu : local.image_ids.nongpu,
+          length(regexall("A1", each.value.shape)) > 0 ? local.image_ids.aarch64 : local.image_ids.x86_64,
+          [for k, v in local.parsed_images : k
+            if length(regexall(v.os, each.value.os)) > 0
+            && trimprefix(v.os_version, each.value.os_version) != v.os_version
+          ],
+      ]...)), 0))
+    }
+  }
+
+  ssh_public_key = (var.ssh_public_key != "") ? var.ssh_public_key : (var.ssh_public_key_path != "none") ? file(var.ssh_public_key_path) : ""
+
+  lifecycle { # prevent resources changes for changed fields
+    ignore_changes = [
+      kubernetes_version, # e.g. if changed as part of an upgrade
+      name, defined_tags, freeform_tags,
+      node_metadata["user_data"],               # templated cloud-init
+      node_config_details["placement_configs"], # dynamic placement configs
+      node_source_details,                      # dynamic image lookup
+    ]
+  }
+
+  dynamic "initial_node_labels" {
+    for_each = merge(var.node_labels, each.value.node_labels)
+    content {
+      key   = initial_node_labels.key
+      value = initial_node_labels.value
+    }
+  }
+}
```

Comment:
We have a nodepool in this module and in the OKE module. How does this work?

Reply:
Implementation in this module is very similar, but with additional functionality consistent with other modes in this sub-module. It can eventually deprecate nodepools.